### ISSUE TITLE
[release/2.14] huggingface DebertaV2ForQuestionAnswering inference amp_fp16 eager performance regression (~100x)

### ISSUE THREAD

**Issue by bjarzemb** [intel/torch-xpu-ops#4928]
### 🐛 Describe the bug

Issue found during pre RC test with torch 2.14.0.dev20260805+xpu. 

There is performance regression on Windows with Intel XPU (Arc B580). Batch size used: 2

Testcase | Model | Inductor Perf status | Inductor Perf diff | Inductor PT2.13 latency [ms] | Inductor PT2.14 latency [ms] | Eager Perf status | Eager Perf diff | Eager PT2.13 latency [ms] | Eager PT2.14 latency [ms]
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
huggingface_performance_inference_amp-fp16 | DebertaV2ForQuestionAnswering | PASS | -2.70% | 44.57 | 43.36 | FAIL | 10898.30% | 70.84 | 7791.14

### Versions

<details>
<summary>Click to expand traceback</summary>
PyTorch version: 2.14.0.dev20260805+xpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro (10.0.26200 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:36:12) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26200-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20260100
Intel GPU driver version:
* 32.0.101.8864 (20260717000000.******+***)
Intel GPU models onboard:
* Intel(R) Arc(TM) B580 Graphics
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Arc(TM) B580 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE20B, uuid=86800be2-0000-0000-0300-000000000000, driver_version='1.15.38308', total_memory=11875MB, local_mem_size=128KB, last_level_cache_size=18432KB, max_compute_units=160, memory_clock_rate=0MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1, is_integrated_gpu=0)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Name: 13th Gen Intel(R) Core(TM) i5-13400
Manufacturer: GenuineIntel
Family: 205
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 2500
MaxClockSpeed: 2500
L2CacheSize: 9728
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] dpcpp-cpp-rt==2026.1.0
[pip3] intel-cmplr-lib-rt==2026.1.0
[pip3] intel-cmplr-lib-ur==2026.1.0
[pip3] intel-cmplr-lic-rt==2026.1.0
[pip3] intel-opencl-rt==2026.1.0
[pip3] intel-openmp==2026.1.0
[pip3] intel-pti==1.0.1
[pip3] intel-sycl-rt==2026.1.0
[pip3] mkl==2026.1.0
[pip3] mkl-include==2024.2.0
[pip3] mkl-static==2024.2.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] onemkl-license==2026.1.0
[pip3] onemkl-sycl-blas==2026.1.0
[pip3] onemkl-sycl-dft==2026.1.0
[pip3] onemkl-sycl-lapack==2026.1.0
[pip3] onemkl-sycl-rng==2026.1.0
[pip3] onemkl-sycl-sparse==2026.1.0
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.1.16
[pip3] onnxscript==0.6.2
[pip3] optree==0.13.0
[pip3] tbb==2023.1.0
[pip3] tbb-devel==2021.13.1
[pip3] tcmlib==1.5.0
[pip3] torch==2.14.0.dev20260805+xpu
[pip3] torchaudio==2.11.0.dev20260805+xpu
[pip3] torchvision==0.29.0.dev20260805+xpu
[pip3] triton-xpu==3.8.0+git1e2d42a0
[pip3] umf==1.1.0
[conda] dpcpp-cpp-rt                 2026.1.0                pypi_0              pypi
[conda] intel-cmplr-lib-rt           2026.1.0                pypi_0              pypi
[conda] intel-cmplr-lib-ur           2026.1.0                pypi_0              pypi
[conda] intel-cmplr-lic-rt           2026.1.0                pypi_0              pypi
[conda] intel-opencl-rt              2026.1.0                pypi_0              pypi
[conda] intel-openmp                 2026.1.0                pypi_0              pypi
[conda] intel-pti                    1.0.1                   pypi_0              pypi
[conda] intel-sycl-rt                2026.1.0                pypi_0              pypi
[conda] mkl                          2026.1.0                pypi_0              pypi
[conda] mkl-include                  2024.2.0                pypi_0              pypi
[conda] mkl-static                   2024.2.0                pypi_0              pypi
[conda] numpy                        1.26.4                  pypi_0              pypi
[conda] onemkl-license               2026.1.0                pypi_0              pypi
[conda] onemkl-sycl-blas             2026.1.0                pypi_0              pypi
[conda] onemkl-sycl-dft              2026.1.0                pypi_0              pypi
[conda] onemkl-sycl-lapack           2026.1.0                pypi_0              pypi
[conda] onemkl-sycl-rng              2026.1.0                pypi_0              pypi
[conda] onemkl-sycl-sparse           2026.1.0                pypi_0              pypi
[conda] optree                       0.13.0                  pypi_0              pypi
[conda] tbb                          2023.1.0                pypi_0              pypi
[conda] tbb-devel                    2021.13.1               pypi_0              pypi
[conda] tcmlib                       1.5.0                   pypi_0              pypi
[conda] torch                        2.14.0.dev20260805+xpu  pypi_0              pypi
[conda] torchaudio                   2.11.0.dev20260805+xpu  pypi_0              pypi
[conda] torchvision                  0.29.0.dev20260805+xpu  pypi_0              pypi
[conda] triton-xpu                   3.8.0+git1e2d42a0       pypi_0              pypi
[conda] umf                          1.1.0                   pypi_0              pypi

</details>

**Comment by fmazurx** [comment]
# Eager Latency Regression Bisection Report

## Summary

A bisection analysis was performed to identify the build in which the Eager Latency regression was introduced and to determine whether there were additional performance degradations afterward.

The investigation identified **two distinct regression points**:

1. A major latency regression introduced in **2.14.0.dev20260723+xpu**, where latency increased from approximately **72 ms** to **1937 ms**.
2. A secondary latency regression introduced in **2.14.0.dev20260728+xpu**, where latency increased further from approximately **1945 ms** to **2314 ms**.

---

## Initial Bisection

### Known Good / Known Bad Builds

| Version | Eager Latency |
|----------|-------------:|
| 2.14.0.dev20260720+xpu | 71.59863862 |
| 2.14.0.dev20260804+xpu | 2310.820692 |

### Bisection Results

| Version | Eager Latency |
|----------|-------------:|
| 2.14.0.dev20260720+xpu | 71.59863862 |
| 2.14.0.dev20260721+xpu | 71.61525835 |
| 2.14.0.dev20260722+xpu | 71.60429285 |
| 2.14.0.dev20260723+xpu | 1936.911203 |
| 2.14.0.dev20260727+xpu | 1944.774495 |
| 2.14.0.dev20260804+xpu | 2310.820692 |

### Finding

The results show that:

- **2.14.0.dev20260720+xpu**
- **2.14.0.dev20260721+xpu**
- **2.14.0.dev20260722+xpu**

all exhibit normal latency of approximately **72 ms**.

Starting with:

- **2.14.0.dev20260723+xpu**

latency increases dramatically to approximately **1937 ms** and remains elevated in all later tested builds.

### Conclusion

**2.14.0.dev20260723+xpu is the first known bad version.**

The primary latency regression was introduced between:

- **2.14.0.dev20260722+xpu** (≈72 ms)
- **2.14.0.dev20260723+xpu** (≈1937 ms)

---

## Secondary Bisection

After identifying the first regression, a second bisection was performed to determine when latency increased further from approximately **1.9 s** to **2.3 s**.

### Bisection Results

| Version | Eager Latency |
|----------|-------------:|
| 2.14.0.dev20260727+xpu | 1944.774495 |
| 2.14.0.dev20260728+xpu | 2313.679710 |
| 2.14.0.dev20260729+xpu | 2310.426632 |
| 2.14.0.dev20260731+xpu | 2309.250344 |
| 2.14.0.dev20260804+xpu | 2310.820692 |

### Finding

The results show that:

- **2.14.0.dev20260727+xpu** still exhibits latency around **1945 ms**.
- **2.14.0.dev20260728+xpu** already exhibits latency around **2314 ms**.
- All later tested builds remain at approximately **2310 ms**.

### Conclusion

**2.14.0.dev20260728+xpu is the first known bad version for the secondary regression.**

The additional latency increase was introduced between:

- **2.14.0.dev20260727+xpu** (≈1945 ms)
- **2.14.0.dev20260728+xpu** (≈2314 ms)

---

## Final Conclusions

Two regression points were identified:

| Regression | Last Good Version | First Bad Version | Latency Change |
|------------|-------------------|-------------------|----------------|
| Primary regression | 2.14.0.dev20260722+xpu | 2.14.0.dev20260723+xpu | ~72 ms → ~1937 ms |
| Secondary regression | 2.14.0.dev20260727+xpu | 2.14.0.dev20260728+xpu | ~1945 ms → ~2314 ms |

### Key Findings

- The major Eager Latency regression was introduced in **2.14.0.dev20260723+xpu**.
- A second performance degradation was introduced in **2.14.0.dev20260728+xpu**.
- The cumulative effect increased latency from approximately **72 ms** in the good builds to approximately **2310 ms** in later builds.

**Comment by PatrykWilczewski** [comment]
Root cause: pytorch#189999, but the change itself is correct - it removes a redundant Half-Float temporary before softmax on XPU and is faster in isolation. It regressed this benchmark only indirectly: the config already reserves ~12GB on an 11876MB B580, and dropping one ~50MB temporary per layer shifted the allocator's segment layout so the hot attention/FFN weights became host-resident instead of cold cached blocks. On Windows an allocation past device capacity doesn't fail (the driver silently backs it with system memory), so the large fp32->fp16 autocast copies then run at ~4-7GB/s instead of ~370GB/s - eager 72.6ms -> 1429ms.
Two options:
1. Default 'expandable_segments' ON for XPU on Windows - reserved then tracks live memory and stays under device capacity; measured eager 1429 -> 67.5 ms (better than the pre-#189999 76ms). Riskier: it makes an opt-in allocator feature the default, so it needs broader testing, and it already exposed a latent segfault on oversized allocations that has to be fixed alongside it.
2. Make the XPU allocator refuse to reserve past device capacity - so this surfaces as a clear OOM isntead of a silent ~20x slowdown. Safer to reason about, but turns some currently-working (slow) workloads into hard failures.

**Comment by PatrykWilczewski** [comment]
Since option 1 can cause various side effects by changing the default behavior, I decieded on a different solution. I was intrigued by the fact that the model doesn't actually requier 12GB of memory, but only a fraction of it (with room for the caching allocator). It turned out that the benchmark was making an unnecessary copy of the model, doubling the required space. More details here: https://github.com/pytorch/pytorch/pull/194575

**Comment by PatrykWilczewski** [comment]
pr 194575 not needed anymore, fixed by https://github.com/pytorch/pytorch/pull/196178
